    Department Of Computer Science
    COMP4381, SP.TOP: DATA SCIENCE AND ANALYTICS
    Dr. Hussein Soboh
    COMP4381 | Section 1 


## <div align=center> Assignment #5 </div>
<div align=center><b>Analyis of a UK-based online retail transactions </b></div>
<div align=center>Occurring between 01/12/2009 and 09/12/2011</div>

<a href="https://www.kaggle.com/datasets/mashlyn/online-retail-ii-uci">https://www.kaggle.com/datasets/mashlyn/online-retail-ii-uci</a>

    Prepeared by: Sondos Aabed   
    ID: 1190652
<a id='intro'></a>
<hr>

**Dear Online Retail company representative,**

The follwing report is written to inform you about the analysis results condocted on the data collected between 01/12/2009 and 09/12/2011 from your UK-based online retail transactions. 

In this analysis report, several key aspects based on your requirements provided earlier are addressed:

1. **Top 5 invoices based on their Total sales**
2. **Top 5 customers based on their Total buying**
3. **Top 5 Selling Items during that Period**
4. **Top 20 Pairs, Most Frequently Purchased Item Pairs by Customers**
5. **Recommendations for a New Promotion Targeting Items with Low Sales Volume**

Please note that this report follows the follwing structure, and you are able to get back up within the notebook.

- Your question and its corresponding number.
- Code cell(s) implementing the necessary calculations and data manipulations. (as a non-technincal reveiwer you can look at the results and the onterpretation)
- Markdown cell presenting the results obtained and providing interpretation.
- You can also check the table of contents, list of figures, and list of tables.

For your information, a random sample of 200k rows from the dataset using (1190652) as the seed to ensure reproducibility.

Insights of this analysis will help in strategic decisions for optimizing sales and customer satisfaction.

Should you have any further queries or require additional clarifications, please do not hesitate to reach out.

Best regards,

<i>Sondos Aabed,</i>

<hr>

#### <b>Table of Contents</b>
- <b>Introduction</b>
- <b>Objectives</b>
- <b>Software Listenings</b>
- <b>Dataset Analysis</b>
    - Load and Sample Data
    - Dataset Initial Exploration
    - Date and total price handling
    - Visualization
- <b>Conclusion</b>
- <b>Refrences</b>

#### <b>List of Figures</b>
- Figure 0: 
- Figure 1: Number of Invoices by Year
- Figure 2: Number of Invoices Over Months
- Figure 3: Word Cloud of Product Descriptions

#### <b>List of Tables</b>
- Table 0: Software Litening
- Table 1: Standard Error of Means for numerical columns
- Table 2: Dataset Head
- Table 3: Dataset Tail
- Table 4: Dataset Numerical Describtion
- Table 5: Missing Values By Percentage 
- Table 6: Dataset Columns Data types
- Table 7: Number of uniques in the datasets
- Table 8: Is Na for datasets

<hr>

#### <b>Introduction</b>

<hr>

#### <b>Objectives</b>
[Back](#intro)


<hr>

#### <b>Software Listenings</b>
[Back](#intro)

The followning table shows the software tools and versions used in this experimnet.

Table 0: Software Listenings

|Tool | Version |
|-----|---------|
|Python|3.12.2|
|Numpy|1.26.4|
|Matplotlib|3.8.2|
|Pandas|2.2.1|
|Visual Studio Code |Updated|
|Git & github|[Repo.](https://github.com/sondosaabed/SP.TOP-Data-Science-and-Analytics/blob/main/Assignments/A4/1190652_A4.ipynb)|

<hr>

#### <b>Dataset Analysis</b>
[Back](#intro)

- Neccessary Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import display
import seaborn as sns
from wordcloud import WordCloud

><b>Loading Data and Sampling</b>

- Read a 200k sample with 1190652 random seed, and calculate sample error

In [5]:
def read_sample(sample_size = 200000, path='./online_retail_II.csv', seed = 1190652):
    """
    Read a csv file into dataframe and take a sample
    Args:
        path (str): the path of the dataset takes a deafult value (relative path)
        seed (int): the seed for randomness so each time it have the same random selcetd
    Returns:
        sample (pd.Dataframe): sample taken from the read df with specific seed
    """
    df = pd.read_csv(path)
    sample = df.sample(sample_size, random_state=seed)
    return sample

In [6]:
sample = read_sample()
sem = pd.DataFrame(sample.sem(numeric_only= True), columns=['SEM'])
print("Table 1: Standard Error of Means for numerical columns:")
display(sem)

(1067371, 8)
Table 1: Standard Error of Means for numerical columns:


SEM
Quantity     0.413771
Price        0.287015
Customer ID  4.313394

A sample of 200k is taken from 1,067,371 population. 


> <b>Dataset Exploration</b>

- Get the missing Values percentages

In [ ]:
def missing_values(df):
    """
    This is to get the percetages of missing data
    Args:
        df (pd.Dataframe): contains the data
    Returns:
        missing_percetanges(pd.Dataframe): contains Column,	Counts, and	Percentage
            of the missing values for eah colmn
    """
    missing_count = df.isnull().sum()
    missing_percetanges = pd.DataFrame({
        'Column': missing_count.index,
        'Counts': missing_count.values,
        'Percentage': missing_count.values / len(df) * 100  
    })
    return  missing_percetanges

- Explore the data including: head, tail, number of uniques, number of NA, statistical describtion for numerical, and data types

In [ ]:
def explore_sample(sample):
    """
    Exploring a dataset sample
    Args:
        sample (pd.Dataframe): the dataset sample to explore.
    Returns:
        results (dict): containing results of each exploration with the title as key
    """
    head = pd.DataFrame(sample.head())
    tail = pd.DataFrame(sample.tail())
    nunique = pd.DataFrame(sample.nunique(), columns=["#_of_Unique"])
    describe = pd.DataFrame(sample.describe())
    isna =  pd.DataFrame(sample.isna().sum(), columns=["#_of_IsNa"])
    dtypes =  pd.DataFrame(sample.dtypes, columns=["Datatype"])
    results = {
        'Table 2: Dataset Head:':head,
        'Table 3: Dataset Tail:':tail,
        'Table 4: Dataset Numerical Describtion: ':describe,
        'Table 5: Missing Values By Percentage': missing_values(sample), 
        'Table 6: Dataset Columns Data types: ':dtypes,
        'Table 7: Number of uniques in the datasets:':nunique,
        'Table 8: Is Na for datasets':isna}
    return results

- Print the dataframes of exploration results.

In [ ]:
def print_sample_exploration(results):
    """
    Prints a beautufil display of each of the exploration dataframe
    Args:
        results (dict): contains exploration outputs with the title as key
    Returns:
        nothing
    """
    for operation, dataframe in results.items():
        print(f"{operation}")
        if operation == 'Missing Values':
            print("Total Sum of Missing Percetange: ", dataframe['Percentage'].sum())
        display(dataframe)

In [ ]:
results = explore_sample(sample)
print_sample_exploration(results)

> <b>Handle Date and Get total Price per inovice</b>

[Back](#intro)

In [ ]:
def handle_date(sample):
    """ 
    Handles date high cardenality
    Args: 
        sample (pd.Dataframe): contains datset
    Return:
        (pd.Dataframe) handeled with day, month, year and time columns added
    """    
    sample['InvoiceDate'] = pd.to_datetime(sample['InvoiceDate'])
    sample['InvoiceYear'] = sample['InvoiceDate'].dt.year
    sample['InvoiceMonth'] = sample['InvoiceDate'].dt.month
    sample['InvoiceDay'] = sample['InvoiceDate'].dt.day
    sample['InvoiceTime'] = sample['InvoiceDate'].dt.time
    sample = sample.drop('InvoiceDate', axis=1)
    return sample

In [ ]:
def get_total(row):
    """
    Calcualtes the total prices by multiplication and considering quantity
    Args:
        row of the dataframe
    Returns:
        Total price  that is calculated by multiplication of Quantity*Price 
    """
    return row["Quantity"] * row["Price"]

In [ ]:
sample = handle_date(sample)
sample['total_price'] = sample.apply(get_total, axis=1)
sample.head()

> <b>Visualization</b>

[Back](#intro)

In [ ]:
sample['InvoiceYear'].value_counts().sort_index().plot(kind='bar')
plt.title('Figure 1: Number of Invoices by Year')
plt.xlabel('Year')
plt.ylabel('Number of Invoices')

In [ ]:
sample.groupby('InvoiceMonth')['Invoice'].count().plot(marker='o')
plt.title('Figure 2: Number of Invoices Over Months')
plt.xlabel('Month')
plt.ylabel('Number of Invoices')

In [ ]:
words = ' '.join(sample['Description'].dropna())
wordcloud = WordCloud(background_color='white').generate(words)
plt.imshow(wordcloud)
plt.title('Figure 3: Word Cloud of Product Descriptions')
plt.axis('off')

In [ ]:
hourly_sales = sample.groupby('InvoiceTime')['total_price'].sum()
peak_sales_hour = hourly_sales.idxmax()
print(f"The peak sales hour is at {peak_sales_hour} o'clock.")

><b>Data Insights</b>

[Back](#intro)

1. Determine the total sales per invoice and identify the top 5 invoices.

In [ ]:
print("The top 5 total Sales and their invovice:")
sample.groupby(sample['Invoice'])['total_price'].sum().nlargest(5)

2. Compute the total sales per customer and list the top 5 customers.

In [ ]:
print("The top 5 buyers by Id:")
sample.groupby(sample['Customer ID'])['total_price'].sum().nlargest(5)

3. Identify the top 5 selling items during this period.

In [ ]:
print("The top 5 selling items:")
sample.groupby(sample['Description'])['total_price'].sum().nlargest(5)

4. Determine the most frequently purchased item pairs by customers and display the top 20 pairs.

5. The retail store aims to introduce a new promotion targeting items with low sales volume. This promotion entails offering a discounted price for a pair of two different items, with one item being among the top 10 best sellers and the other from the bottom 10 sellers. Customers who purchase these two items together will qualify for the special pricing. Your task is to recommend 10 pairs of items to kickstart this campaign.

<hr>

#### <b>Conclusion</b>
[Back](#intro)

<hr>

#### <b>Refrences</b>
[Back](#intro)

- [1] https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.nlargest.html
- [2] https://stackoverflow.com/questions/26873127/show-dataframe-as-table-in-ipython-notebook
<hr>